opds2_import.py

```diff
@@ -0,0 +1,830 @@
+                encode(subject), encode(subject_metadata_list)
+            )
+        )
+
+        return subject_metadata_list
+
+    def _extract_contributors(self, contributor):
+        """Extract a list of ContributorData objects from the webpub-manifest-parser's contributor.
+
+        :param contributor: Parsed contributor object
+        :type contributor: Union[
+            str,
+            List[Union[str, core_ast.Contributor]],
+            core_ast.Contributor
+        ]
+
+        :return: List of contributors metadata
+        :rtype: List[ContributorData]
+        """
+        self._logger.debug(
+            u"Started extracting contributor metadata from {0}".format(
+                encode(contributor)
+            )
+        )
+
+        contributor_metadata_list = []
+
+        if is_string(contributor):
+            contributor_metadata = ContributorData(
+                sort_name=None,
+                display_name=contributor,
+                family_name=None,
+                wikipedia_name=None,
+                roles=None,
+            )
+
+            contributor_metadata_list.append(contributor_metadata)
+        elif isinstance(contributor, core_ast.Contributor):
+            contributor_metadata = ContributorData(
+                sort_name=contributor.sort_as,
+                display_name=contributor.name,
+                family_name=None,
+                wikipedia_name=None,
+                roles=contributor.roles if contributor.roles else None,
+            )
+
+            contributor_metadata_list.append(contributor_metadata)
+        elif isinstance(contributor, list):
+            for inner_contributor in contributor:
+                inner_contributor_metadata_list = self._extract_contributors(
+                    inner_contributor
+                )
+
+                contributor_metadata_list.extend(inner_contributor_metadata_list)
+
+        self._logger.debug(
+            u"Finished extracting contributor metadata from {0}: {1}".format(
+                encode(contributor), encode(contributor_metadata_list)
+            )
+        )
+
+        return contributor_metadata_list
+
+    def _extract_link(self, link, feed_self_url, default_link_rel=None):
+        """Extract a LinkData object from webpub-manifest-parser's link.
+
+        :param link: webpub-manifest-parser's link
+        :type link: ast_core.Link
+
+        :param feed_self_url: Feed's self URL
+        :type feed_self_url: str
+
+        :param default_link_rel: Default link's relation
+        :type default_link_rel: Optional[str]
+
+        :return: Link metadata
+        :rtype: LinkData
+        """
+        self._logger.debug(
+            u"Started extracting link metadata from {0}".format(encode(link))
+        )
+
+        # FIXME: It seems that OPDS 2.0 spec doesn't contain information about rights so we use the default one.
+        rights_uri = RightsStatus.rights_uri_from_string("")
```

Comment:
If it helps, `rights_uri` is only relevant for actual copies of the book -- basically when the link relation is http://opds-spec.org/acquisition/open-access. In that case the default rights_uri can be GENERIC_OPEN_ACCESS. Maybe http://opds-spec.org/acquisition/ is also relevant for self-hosted books that are not open-access; in that case the default rights_uri can be IN_COPYRIGHT.